Add deCONZ component

Kane610 opened the pull request:
## Description:
Component supporting Deconz (lights, sensors) rest API and web sockets for push notification.

The Implementation is more or less code complete.

The Component supports all devices (on a theoretical level) supported by deCONZ. There are lists of verified devices in the documentation.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#3967
**Pull request in [home-assistant-polymer](https://github.com/home-assistant/home-assistant-polymer) with new logo for configurator:** home-assistant/home-assistant-polymer#689

## Example entry for `configuration.yaml` (if applicable):
```yaml
deconz:
  host: 192.168.0.90
  api_key: 1234567890
```

If you don't have a API key and want to generate one use this config and then follow the steps in the GUI:
```yaml
deconz:
  host: 192.168.0.90
```
The component will automatically create an API key and put it into deconz.conf.

If you prefer you can use auto discovery component instead and follow instructions in GUI.

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Kane610 committed:
Base implementation of component, no sensors yet

Kane610 committed:
Added senor files

Kane610 committed:
First fully working chain of sensors and binary sensors going from hardware in to hass

Kane610 committed:
Clean up

Kane610 committed:
Clean up

Kane610 committed:
Added light platform

Kane610 committed:
Turning lights on and off and set brightness now works

Kane610 committed:
Pydeconz is now a proper pypi package
Stop sessions when Home Assistant is shutting down
Use a simpler websocket client

Kane610 committed:
Updated pydocstrings
Followed recommendations from pylint and flake8

Kane610 committed:
Clean up

Kane610 committed:
Updated requirements_all.txt

Kane610 committed:
Updated Codeowners to include deconz.py
Also re-added the Axis component since it had gotten removed

Kane610 committed:
Bump requirement

Kane610 committed:
Bumped to v2
Reran script/gen_requirements

Kane610 committed:
Removed global DECONZ since it wasn't relevant any more

Kane610 committed:
Username and password is only relevant in the context of getting a API key

Kane610 committed:
Add support for additional sensors

Kane610 committed:
Added support for groups

Kane610 committed:
Moved import of component library to inside of methods

Kane610 committed:
Moved the need for device id to library

Kane610 committed:
Bump pydeconz to v5

Kane610 committed:
Add support for colored lights

Kane610 committed:
Pylint and flake8 import improvements

Kane610 committed:
DATA_DECONZ TO DECONZ_DATA

Kane610 committed:
Add support for transition time

Kane610 committed:
Add support for flash

Kane610 committed:
Bump to v7

Kane610 committed:
ZHASwitch devices will now only generate events by default, instead of being a sensor entity

Kane610 committed:
Clean up

Kane610 committed:
Add battery sensor when device signals through an event

Kane610 committed:
Third-party library communicates with service

Kane610 committed:
Add support for effect colorloop

Kane610 committed:
Bump to pydeconz v8

Kane610 committed:
Same domain everywhere

Kane610 committed:
Clean up

Kane610 committed:
Updated requirements_all

Kane610 committed:
Generated API key will now be stored in a config file

Kane610 committed:
Change battery sensor to register to callback since library now supports multiple callbacks
Move DeconzEvent to hub
Bump to v9

Kane610 committed:
Improve entity attributes

Kane610 committed:
Change end of battery name to battery level
No need for static icon variable when using battery level helper

Kane610 committed:
Bump requirement to v10

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+DOMAIN = 'deconz'
+
+DECONZ_DATA = 'deconz_data'
+CONFIG_FILE = 'deconz.conf'
```
I think the path to the config file should be configurable.

Kane610 replied:
This is just the name, the read_json, load_json gets the full path from HASS. From my understanding this is standard behaviour.

tinloaf replied:
As, I see what that file is for.

balloob replied:
If it's not meant for users to manually read/write it, prefix the name with a `.`

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_PORT, default=80): cv.port,
+        vol.Optional(TYPE_AS_EVENT, default=['ZHASwitch']):
```
What is this supposed to do? It looks strange, and it's not documented in the accompanying documentation PR.

Kane610 replied:
Normal behaviour for momentary sensors are to be events, I'd like to have an opening in getting it as a sensor entity instead by disabling it. I'm just keeping it configurable.

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+DECONZ_DATA = 'deconz_data'
+CONFIG_FILE = 'deconz.conf'
+
+TYPE_AS_EVENT = 'type_as_event'
```
For naming reasons, this should probably also start with CONF_.

Kane610 replied:
Sure, good catch!

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+def async_setup(hass, config):
+    """Setup services for Deconz."""
+    deconz_config = config[DOMAIN]
+    config_file = load_json(hass.config.path(CONFIG_FILE))
```
Not sure, but: Since this is a coroutine, it runs in the event loop, and `load_json` probably blocks. Is there an asynchroneous way of reading the JSON?

Kane610 replied:
There is no similar async version but from what I could see in Discord, Balloob didn't think this was an issue.

tinloaf replied:
Okay, great! :+1:

balloob replied:
uuuh what. This is a really big issue. NEVER do I/O inside the event loop.

It is not a big issue that no async version exists, because one does never do I/O inside the event loop. Instead, do `yield from hass.async_add_job(load_json, hass.config.path(CONFIG_FILE))`. Although, if you are not familiar with how asyncio works, it is preferred that you don't write async components. Write sync ones instead.

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+        Store API key in deconz.conf.
+        """
+        from pydeconz.utils import get_api_key
+        deconz_config[CONF_USERNAME] = call.data.get(CONF_USERNAME, 'delight')
```
I don't think that the default values for username / password should be hard coded here, but rather added as default value to the config schema. Nobody's going to look for the default value here. ;)

Kane610 replied:
I guess I could create a second config schema for the service? Since the username and password is only relevant for the service to generate the API key. Not sure what is best :)

tinloaf replied:
Oh, I just now realize that this is a service! Why do you make this a service? The component cannot work before that service was called / the API key was generated, can it? So how about this workflow:
* On startup, check if an API key was generated
  * If so, initialize the component
  * If not, retrieve username / password from the config, generate a new API key
    * If that worked, initialize the component
    * If it didn't, fail the initialization

Kane610 replied:
I had it like that initially, but then I thought that I don't want the user to need to remove the username and password afterwords. Originally I created a notification in the GUI for the user to write down and add to the configuration.yaml. This is the opposite in that after initially specifying the component you are not required to add any additional configuration for the component.

But in the end I guess it is a matter of taste.

Kane610 replied:
I'm gonna think a bit more about this, you have some valid points!

Kane610 replied:
I changed it to more as you suggested

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+        deconz_config[CONF_USERNAME] = call.data.get(CONF_USERNAME, 'delight')
+        deconz_config[CONF_PASSWORD] = call.data.get(CONF_PASSWORD, 'delight')
+        api_key = yield from get_api_key(hass.loop, **deconz_config)
+        if not save_json(
```
Same as above - is there an async way of doing this?

Kane610 replied:
See previous comment.

tinloaf replied:
:+1:

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+    else:
+        hass.services.async_register(
+            DOMAIN, 'generate_api_key', generate_api_key)
+        _LOGGER.warning("Deconz needs API key to set up session. See docs.")
```
If the component can't do anything without an API key, this should probably be an error, not just a warning (i.e., return `False` here)

Kane610 replied:
An error is a bit extreme since you still intentionally have to enable the component. Which means that if you don't specify the API key you have the intent of generating one using the service which will only load when the API key is missing.

tinloaf replied:
Yes, I agree that a failure would be too hard if you have to call the `generate_api_key` service manually. But with the workflow from above, that should not be necessary, should it?

Kane610 replied:
Absolutely. See the other post for my reasoning.

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+                                                     DOMAIN,
+                                                     deconz_config,
+                                                     config))
+    deconz.start()
```
Does this block?

Kane610 replied:
No, it is a asyncio transport class.

tinloaf replied:
:+1:

tinloaf commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class DeconzEvent(object):
+    """When you want signals instead of entities."""
```
I'm afraid I don't understand what that means. :(

Kane610 replied:
I understand this since you only had time to spend on this file ;).

It is a event class that will take a device sensor and generate events in hass instead of creating a sensor entity.

I will improve the pydocstring.

Kane610 committed:
Improve pydocstring for DeconzEvent
Rename TYPE_AS_EVENT to CONF_TYPE_AS_EVENT

amelchio commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,168 @@
+        data = {'on': True}
+
+        if ATTR_BRIGHTNESS in kwargs:
+            data['bri'] = kwargs[ATTR_BRIGHTNESS]
```
You might want to move this down below RGB so one can specify both and have the RGB brightness overridden.

Kane610 replied:
I hear you.

Kane610 committed:
Allow separate brightness to override RGB brightness

Kane610 committed:
Expose device.reachable in entity available property

tinloaf commented:
Regarding the event / entity thing: I think I got it now. I wonder though what the advantage of the "event mode" is over having an entity. If you have an entity for every sensor, you still get an event every time the sensor's state changes (a `STATE_CHANGED` event), so "entity" mode basically has all features of the "event" mode … and then some.

Also, from what I know it's pretty uncommon to not represent physical devices by entities, but by a stream of events instead. I think if my assumption above is correct (that the event mode doesn't give you any "extra features") I would be in favor of just using the "entity mode". Reduces code complexity and makes this component behave more like all other components.

Kane610 commented:
This specific thing I've had a long talk with another author of a Decent component as well as Balloob to get a proper decision. I can agree that some people would prefer entities for all devices, that's why I left the 'undocumented' configuration possibility to alter this from the default behaviour.

tinloaf commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,150 @@
+    @property
+    def unit_of_measurement(self):
+        """Unit of measurement of this entity."""
+        return self._sensor.sensor_unit
```
I think you should translate here into home assistant - standardized units such as `TEMP_CELSIUS` (which is defined as `'C'`, while pydeconz would use `'Celsius'`). The same probably holds for other units?

Kane610 replied:
Good catch! Too quick on that one.

tinloaf commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,92 @@
+        return self._sensor.sensor_icon
+
+    @property
+    def unit_of_measurement(self):
```
See the other sensor file - maybe translation would be benefical here, although I am not sure what unit a binary sensor should have.

Kane610 replied:
No that is true, copy paste issue...

tinloaf commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,92 @@
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        attr = {
+            'battery_level': self._sensor.battery,
```
`ATTR_BATTERY_LEVEL` from `const.py`

Kane610 replied:
Good point!

tinloaf commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,92 @@
+            'manufacturer': self._sensor.manufacturer,
+            'model_number': self._sensor.modelid,
+            'reachable': self._sensor.reachable,
+            'uniqueid': self._sensor.uniqueid,
```
`ATTR_ID` from `const.py` ?

Kane610 replied:
Good point!

tinloaf commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,92 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        attr = {
```
For the attribute names that cannot be imported from `const.py`, I think it would be cleaner to define your own `ATTR_*` constants in the component, so that you have one central place where they are defined and could be changed.

Kane610 replied:
Sure

Kane610 committed:
Bump requirement to 11 (it goes up to 11!)

tinloaf commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,173 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        if self._light.type != 'LightGroup':
```
See binary sensor.

tinloaf commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,150 @@
+        return False
+
+    @property
+    def device_state_attributes(self):
```
See binary sensor

Kane610 committed:
Pylint comment

Kane610 committed:
Binary sensors don't have unit of measurement

tinloaf commented:
My preference for "entity only" is less because of what the user might want, but because of (in my eyes unnecessarily) added code complexity. I mean, you could remove two out of five classes (the event class and the battery class), you would not have the (somewhat strange, I think) "battery" entities, …

Also, breaking the "an phyiscal device is an entity" rule makes your component unusable with some other components, for example the recorder, mqtt statestream, anything that wants to do templating, anything that records historical data, …

But if @balloob is on okay with this, it's probably okay. :wink:

Kane610 commented:
Valid points

Kane610 committed:
Removed service to generate API key in favor of just generating it as a last resort of no API key is specified in configuration.yaml or deconz.conf

Kane610 committed:
Replace clear text to attribute definitions

tinloaf reviewed (dismissed):
I just had another quick discussion with @balloob , and we think that physical sensors should always be represented by an entity, instead of "hiding" them, generating custom events plus a battery sensor. 

We have to try to keep a certain consistency throughout the home assistant code base to ensure maintainability. Since we're not able to see the advantages, we think the added code complexity and different approach from all other sensor platforms is not worth it.

Kane610 commented:
Sure, as long as there isn't a third change back to events later on ;)

tinloaf commented:
Just to make sure: We're really talking about sensors here, not some buttons that are internally called "sensors", are we?

Kane610 commented:
No it is representing remote controls, they're defined ZHASwitch which is a zigbee sensor type

tinloaf commented:
Hum… and these are the only ones that you wouldn't create an entity for, but only generate events from? That makes it a bit more clear. However, I still think everything should just be an entity. However, if that is something you "click", you could think about generating an *additional* `button_clicked` event or so.

Kane610 commented:
Is there a reason to have an event if you have the entity? I'd say it is one or the other.

tinloaf commented:
Well, it's just more intuitive to listen for a `button_clicked` event than for an `state_changed` event, if that state is "the button is pressed now". By the way, would you generate one or two events per click?

Kane610 commented:
One event per click, or rather, the remote gives one event for button pressed down, one for button released, also it can give a third for button press long

tinloaf commented:
Yes, that's what I thought: With a custom event it's easier to distinguish between long / short presses.

Kane610 commented:
Why is that easier? You will have the same information regardless of if it is an event or entity state update

Kane610 committed:
Use more constants

Kane610 commented:
So should I just create an event if the device is a ZHASwitch and also create the entity? And also remove the battery

Kane610 committed:
Bump requirements to v12

Kane610 committed:
Color temp requires xy color support

tinloaf commented:
Now that I know we're talking about remotes, I'm not so 100% sure anymore. :laughing:  Sorry for all the back and forth, and thanks for bearing with me. 

Personally I would still say make everything an entity - you need one for the battery anyways, so that entity might very well also represent the remote. However, I would *also* generate click events (`SHORT_CLICK` / `LONG_CLICK`? Does it support double clicks?), because that's easier to listen for than the `STATE_CHANGED` events from the entity.

However, if you really, really want the event-only solution, please at the least hard-code the sensor types for which to use the "event only" mode. Exposing that as configuration option does not gain us anything (or does it?) and probably confuses users.

Kane610 commented:
From the documentation on how the buttons work
+
 +| Switch code | Description             |
 +|-------------|-------------------------|
 +| 1XXX        | Button #1 up to #8      |
 +| XXX1        | Button hold             |
 +| XXX2        | Button short release    |
 +| XXX3        | Button long release     |
 +
 +Where for example on a Philips Hue Dimmer, 2001 would be holding the dim up button.

So I leave it up to the end user to decide, double clicks should work as well. I'm documenting as I move forward

Kane610 commented:
As I stated initially, I presented the situation to Balloob and had him choose from a few situations :)

Kane610 committed:
Only ZHASwitch should be an event

donnib commented:
@Kane610 @tinloaf so i am ok of having both (both events and entities) but i don't see much usage for an entity for a ZHASwitch which only responds with some code like 1001 or 2001 or whatever code it comes back with because that depends on the manufaterer. That is only what ZHASwitches does sned some codes. The codes doesn't really matter for HASS but for the user they do because 1001 could mean he clicked button up and 1002 could be button down. This information i don't see useful in the history, in the recorder maybe.

For me as a user i think that if you make an entity for a switch i shouldn't spend time to write a template sensore for each switch to get the battery out from the attribute. There should be sensors for battery for each sensor just like @Kane610 has implemented. IMO

It's obvious that cleanest approach is to create entities for all sensors where some sensors will have more entities than one e.g a battery sesnors, this is inline with other components in HASS. Also as an additional bonus add events for sensors would be great IMO.

tinloaf commented:
Having slept again over this, you're probably both right. For some thing that can just be clicked and has no real "state", the entity would have a weird state. So basically: Sorry for all the noise. But I would still think that hard-coding the sensor types that should be events instead of entities is better than offering a hard to understand option to the user. Would that be a compromise? Or is it not easy to come up with a list of sensor types?

Kane610 commented:
I've changed the configuration parameter to be a boolean. So right now the switch will be (default) event + batterysensor else it will be an entity for the device

Kane610 committed:
Bump requirements to v13

Kane610 committed:
Added effect_list property

Kane610 committed:
Add attribute to battery sensor to easy find event id

Kane610 committed:
Bump requirements to v14

Kane610 committed:
Fix hound comment

kroimon commented:
I would love to see scene support for deconz! I think in contrast to the Hue component, there will only be user defined scenes here that the user wants to be able to activate from within Home Assistant.

Kane610 commented:
Do you have experience with deconz scenes? How would you see support exposed in HASS?

kroimon commented:
I think automatically created [Scene](https://home-assistant.io/components/scene/) entities would be great, probably something like the [LiteJet scenes](https://github.com/home-assistant/home-assistant/blob/master/homeassistant/components/scene/litejet.py).

The deconz API allows listing [all existing groups](https://dresden-elektronik.github.io/deconz-rest-doc/groups/#getall) and [their scenes](https://dresden-elektronik.github.io/deconz-rest-doc/scenes/#getall). This is a one-to-one reference to the underlying ZLL scenes.
[Recalling scenes](https://dresden-elektronik.github.io/deconz-rest-doc/scenes/#recall) requires the scene's parent group id and the scene id.

Kane610 commented:
Can you test scene support if I add it?

kroimon commented:
Sure thing! I'll see if I can get a development build running :-)

Kane610 committed:
Bumped requirements_all information to v14

Kane610 committed:
Add service to configure devices on deCONZ

Kane610 committed:
Add initial support for scenes

Kane610 committed:
Bump requirements to v15

Kane610 commented:
@kroimon I've pushed an initial implementation of scene support. Please test it out!

kroimon commented on CODEOWNERS:
```diff
@@ -62,7 +62,9 @@ homeassistant/components/switch/rainmachine.py @bachya
 homeassistant/components/switch/tplink.py @rytilahti
 homeassistant/components/xiaomi_aqara.py @danielhiversen @syssi
 
+homeassistant/components/*/axis.py @kane610
```
Is this part of the deconz PR?

Kane610 replied:
It got overridden in a separate PR so I just add it again here out of pure laziness

kroimon commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,51 @@
+
+
+class DeconzScene(Scene):
+    """Representation of a single LiteJet scene."""
```
Copy&paste error from LiteJet class.

Kane610 replied:
Missed a lot of stuff apparently :)

kroimon commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,51 @@
+    """Representation of a single LiteJet scene."""
+
+    def __init__(self, scene):
+        """Setup sensor and add update callback to get data from websocket."""
```
Add update callback? Copy&paste error?

Kane610 replied:
Update callback ain't necessary here, scenes don't have states. But yeah gotta remove the text.

kroimon commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,51 @@
+    @asyncio.coroutine
+    def async_activate(self, **kwargs):
+        """Activate the scene."""
+        print('scene active kwargs', kwargs)
```
Debug artifact

Kane610 replied:
Yes it is :)

Kane610 committed:
Fix review comments

Kane610 committed:
Python doc string improvement

kroimon commented:
I tested the component with my deconz setup. So far, everything (including scene support) works like a charm!
I also like how light groups are represented as lights so I can change state/color/brightness per group!

kroimon commented:
I have one additional feature request:
Like the hue component, it would be great to have integration with the [configurator component](https://home-assistant.io/components/configurator/) which asks the user to unlock the gateway via 'Settings' -> 'Unlock Gateway' if no API key is given in the configuration or in deconz.conf.

kroimon commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+    if DECONZ_DATA in hass.data:
+        sensors = hass.data[DECONZ_DATA].sensors
+
+    for _, sensor in sensors.items():
```
Should probably be inside of the above `if` statement, otherwise `sensors` will be undefined.

Kane610 replied:
Sensors are initiated as a dict so not an issue.

kroimon replied:
Only if `hass.data[DECONZ_DATA]` could be found (then it should be an instance of `DeconzSession),` otherwise the variable will not be set at all here. So maybe either remove the above `if DECONZ_DATA in hass.data:` completely and rely on it being set, or move the `for` loops into the `if` statement.

Or am I missing something here?

Kane610 replied:
Aah, sure. Yes, my plan is to remove the if statement

kroimon commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,165 @@
+        deconz_config[CONF_API_KEY] = config_file[CONF_API_KEY]
+    else:
+        from pydeconz.utils import get_api_key
+        api_key = yield from get_api_key(hass.loop, **deconz_config)
```
What if this can not get a valid API key because either no username/password was given or the gateway has not been unlocked?
You should guard the following lines against this case, otherwise a `deconz.conf` containing `"api_key": false` will be written.

Kane610 replied:
Sure, should do a verification of the api key

kroimon commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,182 @@
+        lights = hass.data[DECONZ_DATA].lights
+        groups = hass.data[DECONZ_DATA].groups
+
+    for _, light in lights.items():
```
Should probably be inside of the above `if` statement, otherwise `lights` and `groups` will be undefined.

Kane610 replied:
They're created as empty dicts so it will work out as expected.

Kane610 commented:
I can look in to it, good suggestion! My initial thought is that since deCONZ imitates a Hue bridge this would generate falls positives, so I'm not sure this will work to a 100% and if that is the case I'd rather not implement support for it.

kroimon commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,47 @@
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up scenes for the deCONZ platform."""
+    scenes = hass.data[DECONZ_DATA].scenes
```
Guard with `if DECONZ_DATA in hass.data:`?

Kane610 replied:
You're not supposed to get to the platform setup if the initial setup won't work. I'm removing that check on all places :)

kroimon replied:
Oh get it, that's fine.

balloob replied:
@kroimon is right, people could still add it to their config.

Guard for it by checking if `discovery_info` is None:

```python
if discovery_info is None:
    return False
```

The other setup_platform will come through quickly after that.

Kane610 replied:
Im not understanding the why. I'm doing the change but would appreciate a clarification of the reasoning

kroimon commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,167 @@
+    if DECONZ_DATA in hass.data:
+        sensors = hass.data[DECONZ_DATA].sensors
+
+    for _, sensor in sensors.items():
```
`sensors` might be undefined here.

Kane610 replied:
It will be empty at most, so it will work out.

kroimon commented:
What false positives do you think that might generate? Deconz will reply with a 403 error whenever the API key is invalid. This can be the case when someone reset the deconz database or if we don't have a valid API key at all. In those cases, showing a hint to unlock the gateway and press a button to reconfigure should be fine.

Kane610 commented:
Aah, I read as though you meant auto discovery of Deconz, which would be false positive since it identifies as a Hue bridge. If you still expect host address to be specified that is a whole different story indeed.

kroimon commented:
Yes you are right. NetDisco will detect deconz as a Hue bridge right now, triggering the hue component. I'm currently trying to find out if there is a way to differentiate the two, so we could provide auto-discovery of the deconz component. This way, we wouldn't even need the user to specify a host address in the configuration file.

Kane610 committed:
Improve setup and error handling during setup

Kane610 commented:
That would be great!

Kane610 commented:
It could be as easy as identifying that the MAC is not in a series owned by Philips

Kane610 committed:
Changed how to evaluate light features

Kane610 committed:
Remove 'ghost' events by not triggering updates if the signal originates from a config event
Bump requirement to v17

Kane610 committed:
Fix pylint issue by moving scene ownership in to groups in requirement pydeconz
Bump requirement to v18

Kane610 committed:
Added configurator option to register to deCONZ when unlocking gateway through settings
Bump requirement to v20

Kane610 commented:
@kroimon added your unlock gateway request. Seems to work, can you verify?

pietervandersluis commented:
Past weekend, I gave your deconz component a try and I noticed a issue that I had with a custom component. (The custom component wasn't this pull request)

The thing is that the `Xiaomi  Aqara body sensors` only report presence is `True`. Thus if the sensor is triggered the binary sensor is set to `on` and will stay on.

I've noticed this behavior also in your component.

Well, now I feel like a ungrateful person. But I'm really glad that you are building this component.

PS: It wouldn't bother me that much if the behavior remains. A easy workaround would be creating a automation that switches the binary sensor off.

Kane610 commented:
@pietervandersluis 
Does it continuously send out new signals with movement?

There are 3 solutions to that issue

1. Change nothing
2. Change presence detector from entity to event
3. Add a timer in the backend that reverts true to false at timer end when no new presence=true signal has been received

pietervandersluis commented:
@Kane610, yes it sends constantly sends a new signal. The signal is received via the web socket. The only difference in noticeable in the rest api is the `lastupdated` field.

```
pieter$ curl http://172.20.10.69/api/XXXXXX/sensors/10 | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   311  100   311    0     0  22800      0 --:--:-- --:--:-- --:--:-- 23923
{
    "config": {
        "duration": 0,
        "on": true,
        "reachable": true
    },
    "ep": 1,
    "etag": "662509dd57cd98d3f41ee5fe48df3a98",
    "manufacturername": "LUMI",
    "modelid": "lumi.sensor_motion.aq2",
    "name": "Hall Motion",
    "state": {
        "lastupdated": "2017-11-23T21:33:14",
        "presence": true
    },
    "type": "ZHAPresence",
    "uniqueid": "00:15:8d:00:01:65:6d:c1-01-0406"
}
```

The response given via the web socket looks like this:

```
{  
   "e":"changed",
   "id":"10",
   "r":"sensors",
   "state":{  
      "lastupdated":"2017-11-23T21:33:14",
      "presence":true
   },
   "t":"event"
} 
```

manup commented:
Hi, the problem should be solved in latest deCONZ version 2.04.91.

https://github.com/dresden-elektronik/deconz-rest-plugin/commit/e00b544d90bea8d361ab16597db4500f6f38217b

Kane610 commented:
Great @manup, less workarounds for me. Also better when wanting different durations.

balloob commented:
Please remove the deconz support from `light/hue.py`

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+
+    for _, sensor in sensors.items():
+        if sensor.type in DECONZ_BINARY_SENSOR:
+            async_add_devices([DeconzBinarySensor(sensor)], True)
```
Don't call this with 1 entity at a time, just create a list and call it once.

Kane610 replied:
Done

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+    from pydeconz.sensor import DECONZ_BINARY_SENSOR
+    sensors = hass.data[DECONZ_DATA].sensors
+
+    for _, sensor in sensors.items():
```
`for sensor in sensors.values()`

Kane610 replied:
Fixed

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+        self._sensor = sensor
+        self._sensor.register_callback(self._update_callback)
+
+    @callback
```
This is not a callback in the Home Assistant async sense. This only works with `hass.async_add_job` / `hass.add_job`

Kane610 replied:
I double checked with @armills comment on my latest Axis PR and his comment was "The datagram endpoint will be called on the event loop, so it's safe to annotate it as @callback"

armills replied:
Also regarding: https://github.com/home-assistant/home-assistant/pull/10321#discussion_r152920721

Looking through the upstream lib it does look like this should all be scheduled properly on the hass event loop. We should indicate it as such, although I personally don't have any preference between `@callback`, `async_`, or just a docstring note. I'd personally prefer keeping the decorator and adding the `async_` prefix. Even if the decorator is non-functional as it's currently used, it's nice to have the semantic information attached to the function if we ever used it differently.

Breadcrumbs:
https://github.com/Kane610/deconz/blob/1ca6c5fa6dbbe1a796282a13b602c6a85735fb33/pydeconz/websocket.py#L97
https://github.com/Kane610/deconz/blob/1ca6c5fa6dbbe1a796282a13b602c6a85735fb33/pydeconz/__init__.py#L37
https://github.com/Kane610/deconz/blob/1ca6c5fa6dbbe1a796282a13b602c6a85735fb33/pydeconz/__init__.py#L130
https://github.com/Kane610/deconz/blob/1ca6c5fa6dbbe1a796282a13b602c6a85735fb33/pydeconz/deconzdevice.py#L46

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+    def _update_callback(self, reason):
+        """Update the sensor's state, if reason is that state is updated."""
+        if reason['state']:
+            self.async_schedule_update_ha_state()
```
Don't call async methods from a sync context (I doubt deconz schedules your callback on the event loop?)

Kane610 replied:
All callbacks within entities originates from asyncio transport data_received

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+        from pydeconz.sensor import PRESENCE
+        attr = {
+            ATTR_BATTERY_LEVEL: self._sensor.battery,
+            ATTR_FW_VERSION: self._sensor.swversion,
```
Don't include Firmware version, manufacturer version or model id.

Kane610 replied:
Removed

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+            ATTR_FW_VERSION: self._sensor.swversion,
+            ATTR_MANUFACTURER: self._sensor.manufacturer,
+            ATTR_MODEL_ID: self._sensor.modelid,
+            ATTR_UNIQUE_ID: self._sensor.uniqueid,
```
What's this? And why is it relevant for the state?

Kane610 replied:
Removed

balloob commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,91 @@
+            ATTR_UNIQUE_ID: self._sensor.uniqueid,
+        }
+        if self._sensor.type == PRESENCE:
+            attr['dark'] = self._sensor.dark
```
What does this mean?

Kane610 replied:
Dark is extra information from the motion detector whether it is light or dark where it is placed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+DECONZ_DATA = 'deconz_data'
+CONFIG_FILE = 'deconz.conf'
+
+CONF_SWITCH_AS_EVENT = 'switch_as_event'
```
This should not be configurable. 

Buttons can be events, not sure how a switch can be? Switches are on or off and can be toggled.

Kane610 replied:
Switches in this context are remotes like the Hue Dimmer. We had a discussion about this earlier in Discord

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_API_KEY): cv.string,
```
Should one either enter an API_KEY or USERNAME/PASSWORD ? If so, mark it as exclusive.

kroimon replied:
With the new configurator integration, neither API_KEY nor USERNAME/PASSWORD is required, if the user follows the instructions to unlock the gateway. If he still wants to, they should be exclusive, yes. USERNAME/PASSWORD will only be considered if there is no API_KEY given.

After implementing auto-discovery, even the HOST can be made optional.

Kane610 replied:
Username and password are only used to generate API key. I will look into this exclusive thing.

kroimon commented:
I just tested the configurator integration in different scenarios, seems to work as expected!
You might want to flesh out the request and error messages a little, but other than that, I think this feature improves user experience a lot!

As home-assistant/netdisco#156 has just been merged, I feel like the last thing to implement would be support for auto-disovery :-)

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    """Setup services for Deconz."""
+    deconz_config = config[DOMAIN]
+    config_file = load_json(hass.config.path(CONFIG_FILE))
+    descriptions = load_yaml_config_file(
```
No I/O inside async

Kane610 replied:
Fixed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+        """Stop the connections to Deconz on shutdown."""
+        if DECONZ_DATA in hass.data:
+            _LOGGER.info("Stopping Deconz session.")
+            hass.data[DECONZ_DATA].close()
```
Is this doing I/O ?

Kane610 replied:
It calls close to a asyncio transport.

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    @callback
+    def _shutdown(call):  # pylint: disable=unused-argument
+        """Stop the connections to Deconz on shutdown."""
+        if DECONZ_DATA in hass.data:
```
Instead of this if check, only register this callback when `hass.data[DECONZ_DATA]` has been set

Kane610 replied:
Fixed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+        api_key = yield from get_api_key(hass.loop, **deconz_config)
+        if api_key:
+            deconz_config[CONF_API_KEY] = api_key
+            save_json(hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key})
```
No I/O in async context

Kane610 replied:
Fixed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+            deconz_config[CONF_API_KEY] = api_key
+            save_json(hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key})
+        else:
+            hass.async_add_job(
```
Just call `async_request_config`

Kane610 replied:
Done

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    from pydeconz import DeconzSession
+    deconz = DeconzSession(hass.loop, **deconz_config)
+    hass.data[DECONZ_DATA] = deconz
+    result = yield from deconz.load_parameters()
```
Is this doing I/O in the event loop?

Kane610 replied:
Yes it uses aiohttp.clientsession

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    if result is False:
+        _LOGGER.error('Failed to setup deCONZ component')
+        return False
+    hass.async_add_job(discovery.async_load_platform(hass,
```
Why would you add_job these?

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+        if api_key:
+            deconz_config[CONF_API_KEY] = api_key
+            result = yield from _setup_deconz(hass, config, deconz_config)
+            save_json(hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key})
```
No I/O in the event loop

Kane610 replied:
Fixed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+            result = yield from _setup_deconz(hass, config, deconz_config)
+            save_json(hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key})
+            if result:
+                hass.async_add_job(configurator.request_done, request_id)
```
Just call async_request_done

Kane610 replied:
Done

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+            if result:
+                hass.async_add_job(configurator.request_done, request_id)
+                return True
+        hass.async_add_job(
```
Why add job, just call async_notify_errrors

Kane610 replied:
Done

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+        "deCONZ", configuration_callback,
+        description="deCONZ -> Menu -> Settings -> Unlock Gateway",
+        submit_caption="I have unlocked the gateway",
+        link_name='deCONZ platform documentation',
```
`link_name` and `link_url` are deprecated. `description` can now be markdown with as many links as you want.

kroimon replied:
I would suggest adding a link to `http://<deCONZ IP>/edit_system.html` which leads directly to the page with the 'Unlock Gateway' button.

Kane610 replied:
Moved to markdown

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    )
+
+
+class DeconzEvent(object):
```
Why would this be a class instead of a function that just registers a callback and fires events?

Kane610 replied:
This way it fits in directly with how all devices in deconz are represented.

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+        self._event = DOMAIN + '_' + CONF_EVENT
+        self._id = slugify(self._device.name)
+
+    @callback
```
Does deconz call this inside the event loop?

Kane610 replied:
Callback originates from a asyncio transport data_received method

balloob commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,181 @@
+    groups = hass.data[DECONZ_DATA].groups
+
+    for _, light in lights.items():
+        async_add_devices([DeconzLight(light)], True)
```
Collect all entities in a list and call `async_add_devices` once.

Kane610 replied:
Done

balloob commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,181 @@
+
+    for _, group in groups.items():
+        if group.lights:  # Don't create entity for group not containing light
+            async_add_devices([DeconzLight(group)], True)
```
Collect all entities in a list and call `async_add_devices` once.

Kane610 replied:
Done

balloob commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,181 @@
+        """Return the state attributes of the sensor."""
+        if self._light.type != ATTR_LIGHT_GROUP:
+            attr = {
+                ATTR_FW_VERSION: self._light.swversion,
```
Don't include firmware, manufacturer and model

balloob commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,181 @@
+                ATTR_FW_VERSION: self._light.swversion,
+                ATTR_MANUFACTURER: self._light.manufacturer,
+                ATTR_MODEL_ID: self._light.modelid,
+                ATTR_UNIQUE_ID: self._light.uniqueid,
```
Why is this info relevant?

Kane610 replied:
Removed

balloob commented on homeassistant/components/deconz.py:
```diff
@@ -0,0 +1,198 @@
+    if result is False:
+        _LOGGER.error('Failed to setup deCONZ component')
+        return False
+    hass.async_add_job(discovery.async_load_platform(hass,
```
You can satisfy linting by just adding an enter after `(` and indent the parameter list with 4 spaces. Better readable.

Kane610 replied:
Done

balloob commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,48 @@
+    scenes = hass.data[DECONZ_DATA].scenes
+
+    for _, scene in scenes.items():
+        async_add_devices([DeconzScene(scene)])
```
Make 1 list of entities

Kane610 replied:
Done

balloob commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,165 @@
+            if switch_as_event and sensor.type == 'ZHASwitch':
+                DeconzEvent(hass, sensor)
+                if sensor.battery:
+                    async_add_devices([DeconzBattery(sensor)], True)
```
We should not have batteries as standalone sensors. Instead, add a ATTR_BATTERY_LEVEL to the switch.

donnib replied:
@balloob as a user (i was the one requesting this) i like that battery level are separate sensors. I think this makes it much easier to show a list of battery level instead of creating a template sensor for each one to pull the attribute out and show it in the dashboard, is there an easier way to show this ? I noticed for instance that my iphone battery level is added as a separate sensor in HASS, is that not the same thing ?

Kane610 replied:
Balloob, this is what we talked about a few weeks ago on discord. Switch is only for remotes, not a physical wall switch with multiple states.

balloob commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,165 @@
+                if sensor.battery:
+                    async_add_devices([DeconzBattery(sensor)], True)
+            else:
+                async_add_devices([DeconzSensor(sensor)], True)
```
add all entities with 1 call

Kane610 replied:
Done

balloob commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,165 @@
+        """Return the state attributes of the sensor."""
+        attr = {
+            ATTR_BATTERY_LEVEL: self._sensor.battery,
+            ATTR_FW_VERSION: self._sensor.swversion,
```
Don't include firmware, manufacturer etc.

Kane610 replied:
Removed

balloob commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,165 @@
+        """Return the state attributes of the sensor."""
+        attr = {
+            ATTR_EVENT_ID: slugify(self._device.name),
+            ATTR_FW_VERSION: self._device.swversion,
```
Don't include firmware version etc

Kane610 replied:
Removed

balloob commented on homeassistant/components/services.yaml:
```diff
@@ -320,6 +320,17 @@ axis:
         description: What parameter to operate on. [Required]
         example: 'package=VideoMotionDetection'
 
+deconz:
```
We should no longer add more things to this file. Instead rename your component to `deconz/__init__.py` and add a `deconz/services.yaml`

Kane610 replied:
Changed

balloob reviewed (changes requested):
left a bunch of comments

This PR should probably have been split up, first add a component and then add 1 platform. Now it's a) a lot of work to review (and I probably won't get to a review again in a while) and b) the same mistake has been made multiple times and could have been caught in a smaller PR too.

Kane610 committed:
Improve async configurator

Kane610 committed:
No user interaction for deconz.conf

Kane610 committed:
No file management in event loop

Kane610 committed:
Improve readability of load platform

Kane610 committed:
Fewer entity attributes

Kane610 committed:
Use values() instead of items() for dicts where applicable

Kane610 committed:
Do one add devices per platform

Kane610 committed:
Clean up of unused attributes

Kane610 committed:
Make sure that discovery info is not None

Kane610 committed:
Only register configure service and shutdown service when deconz has been setup properly

Kane610 committed:
Move description

Kane610 committed:
Fix lines longer than 80

Kane610 committed:
Moved deconz services to a separate file and moved hub to deconz/__init__.py

Kane610 committed:
Remove option to configure switch as entity

Kane610 committed:
Moved DeconzEvent to sensor since it is only Switch buttonpress that will be sent as event

Kane610 committed:
Added support for automatic discovery of deconz
Thanks to Kroimon for adding support to netdisco

Kane610 committed:
Use markup for configuration description

Kane610 commented:
@kroimon added support for your net disco deconz, see if that is what you requested

Kane610 committed:
Fix coveragerc

Kane610 committed:
Remove deCONZ support from Hue component

Kane610 commented:
@balloob I've fixed all current comments, if you don't have time to review/accept this PR for next release you should probably not bump netdisco either since that might break support for current deconz users through the Hue component.

Kane610 committed:
Improved docstrings and readability

Kane610 committed:
Remove unnecessary extra name for storing in hass.data, using domain instead

kroimon commented:
I just tested auto-discovery and configurator integration. That's exactly what I was hoping for! Great work!

Kane610 commented:
@kroimon excellent! Thanks for fixing netdisco!

Kane610 committed:
Improve readability by renaming all async methods
Bump to v21 - improved async naming on methods

Kane610 committed:
Fix first line not being in imperative mood

Kane610 committed:
Added logo to configurator
Let deconz.conf be visible since it will be the main config for the component after initial setup

Kane610 commented:
@balloob what do you think? Do you need additional reviewing of this component?

tgroeneveldlg commented:
@Kane610 Hi Kane,

I just have a quick question, does it discover the WS port from the deCONZ REST config get ( http://{{ip}}/api/{{username}}/config )

Kane610 commented:
@tgroeneveldlg yes it gets added automatically when you load deconz config.

Kane610 committed:
Removed bridge_type from new unit tests as part of removing deconz support from hue component

tgroeneveldlg commented:
@Kane610 Just want to say that I am very pleased with the results.
I have created a docker image with the PR, and it works like a charm.
Still have to figure out on how to build one for Hassio, as that is where I am running my main HA.

One remark I have is, the color selection seems to be very sluggish. If I do this on a Hue bridge, the colors almost instantly change. Is deCONZ REST less responsive than the Hue bridge ?
I believe that the UI with the status bar graph on top, may cause this somehow.

Excellent job !

Kane610 commented:
@tgroeneveldlg thanks! Great to hear!

There is another guy (@marthoc) who is working with getting deconz to run inside docker. You can check with him if you can help out.

Color selection in hass gui? Or is it from setting it in gui to light updating? Or is it from sending command until frontend is updated?

I don't think there is much that can be improved in sending commands. But if someone says that it can be improved and have a suggestion, Im open for improvements :)

marthoc commented:
@tgroeneveldlg I would be very interested in seeing your Dockerfile, since I've been having lots of issues with getting Conbee/Raspbee reliably detected when deCONZ is running as a containerized app... Actually, are you talking about deCONZ in a container or HASS including this PR in a container...?

tgroeneveldlg commented:
@marthoc you can not run deCONZ in a (regular) docker container. My next step is running deCONZ in a Debian-Hassio (pi) container. But I do need to find some time to build that one.

deCONZ is unfortunately never able to find my ConBee in my container, which is why I run deCONZ on the docker host, which, thank god, was Ubuntu and not Centos.

So it is just the pull request within this container. 
I moved away from my HA in docker, because I can not use LIFX (due to broadcasts never arriving in the container)

marthoc commented:
@tgroeneveldlg Yes building deconz into a container is proving tricky :) . But I think I am quite close to getting it working. Once that happens I can build it into an add-on for Hassio to make integrating quite easy.

tgroeneveldlg commented:
@Kane610 Not sure what is causing it, but basically what I mean is, setting the color via UI using the basic color-wheel in HA.
It seems that if you click and drag around the wheel, the status bar graph is constantly updating, with every pixel I drag the wheel.

Once this get's released in HASSIO, I will do some further troubleshooting and sniff what is transferred when.

Kane610 committed:
Capitalize first letters of Battery Level

tgroeneveldlg commented:
Hi @Kane610 Is there any news on this PR? Do you know what it is waiting for atm?

Kane610 commented:
@tgroeneveldlg no news yet. I'm gonna poke them later this week. I think they've been quite busy with the cloud service

kristianwh commented:
Great work @Kane610 , really looking forward to it!

Kane610 committed:
Properly update state of sensor as well as reachable and battery
Bump dependency to v22

Kane610 committed:
Fix flake8 Multi-line docstring closing quotes should be on a separate line

MartinHjelmare commented:
Please remove the changes to Hue light from your branch.

Kane610 commented:
@MartinHjelmare why? @balloob asked me to remove that explicit support.

MartinHjelmare commented:
Ah, sorry, I didn't see that it was related.

Kane610 commented:
@MartinHjelmare would you mind helping me getting this PR merged?

MartinHjelmare commented:
I'll try to have a look tomorrow.

Kane610 commented:
Great! If you're wondering stuff I'm available here and on discord. Robban1969

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+import asyncio
+import logging
+
+from homeassistant.components.deconz import DOMAIN
```
`... import DOMAIN as DECONZ_DATA`

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+from homeassistant.core import callback
+from homeassistant.util.color import color_RGB_to_xy
+
+DEPENDENCIES = [DOMAIN]
```
`'deconz'`

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+
+DEPENDENCIES = [DOMAIN]
+
+_LOGGER = logging.getLogger(__name__)
```
This isn't used.

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup light for deCONZ component."""
+    if discovery_info is None:
+        return False
```
Just `return`.

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+    if discovery_info is None:
+        return False
+
+    lights = hass.data[DOMAIN].lights
```
```py
lights = hass.data[DECONZ_DATA].lights
```

MartinHjelmare commented on homeassistant/components/binary_sensor/deconz.py:
```diff
@@ -0,0 +1,96 @@
+        return False
+
+    from pydeconz.sensor import DECONZ_BINARY_SENSOR
+    sensors = hass.data[DOMAIN].sensors
```
```py
sensors = hass.data[DECONZ_DATA].sensors
```

MartinHjelmare commented on homeassistant/components/light/deconz.py:
```diff
@@ -0,0 +1,171 @@
+    def __init__(self, light):
+        """Setup light and add update callback to get data from websocket."""
+        self._light = light
+        self._light.register_async_callback(self.async_update_callback)
```
Move registration of callback to `async_added_to_hass`. See above.

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+Support for deCONZ scenes.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/scene.deconz/
```
URL is wrong.

Kane610 replied:
This one is correct?

MartinHjelmare replied:
Yes, you're correct, it's correct! Sorry! :blush:

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+
+DEPENDENCIES = [DOMAIN]
+
+_LOGGER = logging.getLogger(__name__)
```
This isn't used.

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+import asyncio
+import logging
+
+from homeassistant.components.deconz import DOMAIN
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up scenes for deCONZ component."""
+    if discovery_info is None:
+        return False
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+    if discovery_info is None:
+        return False
+
+    scenes = hass.data[DOMAIN].scenes
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+Support for deCONZ sensor.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/sensor/deconz/
```
URL is wrong.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+import asyncio
+import logging
+
+from homeassistant.components.deconz import DOMAIN
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,53 @@
+from homeassistant.components.deconz import DOMAIN
+from homeassistant.components.scene import Scene
+
+DEPENDENCIES = [DOMAIN]
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+
+DEPENDENCIES = [DOMAIN]
+
+_LOGGER = logging.getLogger(__name__)
```
This isn't used.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+from homeassistant.helpers.icon import icon_for_battery_level
+from homeassistant.util import slugify
+
+DEPENDENCIES = [DOMAIN]
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup sensor for deCONZ component."""
+    if discovery_info is None:
+        return False
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+        return False
+
+    from pydeconz.sensor import DECONZ_SENSOR
+    sensors = hass.data[DOMAIN].sensors
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+    def __init__(self, sensor):
+        """Setup sensor and add update callback to get data from websocket."""
+        self._sensor = sensor
+        self._sensor.register_async_callback(self.async_update_callback)
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+        self._name = self._device.name + ' Battery Level'
+        self._device_class = 'battery'
+        self._unit_of_measurement = "%"
+        self._device.register_async_callback(self.async_update_callback)
```
Same as for the other platforms.

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,187 @@
+        self._hass = hass
+        self._device = device
+        self._device.register_async_callback(self.async_update_callback)
+        self._event = DOMAIN + '_' + CONF_EVENT
```
Use new style string formatting and remove the domain constant:
```py
self._event = 'deconz_{}'.format(CONF_EVENT)
```

MartinHjelmare reviewed (changes requested):
In general I think the code flow looks good. One bigger thing is the need to update the deconz library to accept the home assistant aiohttp client session.

See comments below.

Kane610 committed:
Fix martinhjelmares comments

Bump dependency to v23
Use only HASS aiohttp session
Change when to use 'deconz' or domain or deconz data
Clean up unused logger defines
Remove unnecessary return values
Fix faulty references to component documentation
Move callback registration to after entity has been initialized by HASS
Less inception style on pydocs ;)
Simplify loading platforms by using a for loop
Added voluptous schema for service
Yaml file is for deconz only, no need to have the domain present
Remove domain constraint when creating event title

Kane610 commented on homeassistant/components/deconz/__init__.py:
```diff
@@ -0,0 +1,172 @@
+
+DOMAIN = 'deconz'
+
+CONFIG_FILE = 'deconz.conf'
```
There is no easy way for users to get a hold of the API key if needed, so if possible I'd like to leave this as is.

MartinHjelmare replied:
Ok.

Kane610 commented:
@MartinHjelmare thanks a lot for this very detailed review

MartinHjelmare commented on homeassistant/components/sensor/deconz.py:
```diff
@@ -0,0 +1,196 @@
+        self._id = slugify(self._device.name)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
```
Since this class doesn't inherit from `Entity`, you can't use this method, and you don't need to either, since you pass in `hass` when instantiating the event object. So move the registration back to init.

Kane610 replied:
Got carried away with the changes :)

MartinHjelmare commented on homeassistant/components/scene/deconz.py:
```diff
@@ -0,0 +1,50 @@
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
```
This is the default for scenes, so you can remove this property here.

Kane610 replied:
Fixed. Best last day of the year ever if it gets merged today 👍

MartinHjelmare reviewed (commented):
Last comment. Change that and when tests pass, I think this is ready to be merged. :+1:

MartinHjelmare reviewed (approved):
Nice!

Kane610 commented:
@MartinHjelmare what's left for this PR to be merged?

MartinHjelmare commented:
I'm happy but our CLA bot seems to have troubles processing this PR. My guess is the PR simply has too many commits. Try squashing to older commits into one commit if you are comfortable with git. Otherwise you can make a new branch and copy your changes manually and open a new PR and link to this PR.

Kane610 commented:
I will look into how to squash commits tomorrow

Kane610 commented:
@balloob do you have any knowledge about the issue mentioned by MartinHjelmare?

Kane610 commented:
@MartinHjelmare I squashed a few commits, less than a 100 commits is the sweet spot. But it seems to be a bigger issue with travis atm

MartinHjelmare commented:
@Kane610 I've restarted the build. Let's see what happens.

Kane610 commented:
Great @MartinHjelmare ! It is ready for merging 🍾

MartinHjelmare reviewed (approved):
Great!

ScuttleSE commented:
So...I'm not sure I'm doing this correctly, but I have a bug report for this component :)

I have merged this pull request with my current installation of 0.60 and have gotten the deCONZ integration running just fine.

Except for one type of sensor I have; the Aqara Tempterature/Humidity sensor.

When this is added to HA from deCONZ it is added as three sensors. One for temperature, one for humidity and one for airpressure.

They are added as sensor.[name in deconz], sensor.[name in deconz]_2 and sensor.[name in deconz]_3

The problem is that this order is seeminly random. After restarting HA, the sensor that previously was temperature might be the humidity or airpressure sensor. 

Attaching two screenshots to illustrate. The first image is before restarting HA, the second one is after restarting.

![sensor1a](https://user-images.githubusercontent.com/4344284/34556930-b1d38924-f138-11e7-8512-80725fc96d26.PNG)

![sensor1b](https://user-images.githubusercontent.com/4344284/34556946-c956e564-f138-11e7-9e22-e9e884af4a2b.PNG)

Kane610 commented:
@ScuttleSE this is known. It is expected that the user name their devices uniquely.

ScuttleSE commented:
How though? This is one device in deCONZ named temp_1_balcony which in HA gets named sensor.temp_1_balcony, sensor.temp_1_balcony_2 and sensor.temp_1_balcony_3

Kane610 commented:
@ScuttleSE either through the API, which you can use the deconz service in hass for, I think you can also use the new web frontend phoscon to do it. I'm also planning on looking in to doing frontend support later on for deconz

ScuttleSE commented:
But...maybe I am misunderstanding here... :)

I have named the device temp_1_balcony using the phoscon webthing, that's where I added the sensor too, but there is only an option to name the whole device there, not the three "sub-sensors".

When the device gets sucked into HA through the API, it gets added as three separate devices, and HA seems to add the extensions _2 and _3 to two of the sensors at random to be able to differentiate between them

Kane610 commented:
As I said, lets take further discussions in the forum https://community.home-assistant.io/t/conbee-zigbee-stick-and-ha-will-this-work-together/

ScuttleSE commented:
I would if the site weren't in read-only mode :)

Kane610 commented:
I don't understand how that could be, have you posted earlier in the forums?

MartinHjelmare commented:
Please open an issue if you suspect a bug. If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Merged PRs should not be used for support, bug reports or to discuss the implementation. Thanks!
